crates/iceberg/src/spec/table_metadata_builder.rs

```diff
@@ -0,0 +1,2074 @@
+        if spec_id == Self::LAST_ADDED {
+            spec_id = self.last_added_spec_id.ok_or_else(|| {
+                Error::new(
+                    ErrorKind::DataInvalid,
+                    "Cannot set default partition spec to last added spec: no spec has been added.",
+                )
+            })?;
+        }
+
+        if self.metadata.default_spec.spec_id() == spec_id {
+            return Ok(self);
+        }
+
+        if !self.metadata.partition_specs.contains_key(&spec_id) {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!("Cannot set default partition spec to unknown spec with id: '{spec_id}'",),
+            ));
+        }
+
+        let schemaless_spec =
+            self.metadata
+                .partition_specs
+                .get(&spec_id)
+                .ok_or_else(|| {
+                    Error::new(
+                ErrorKind::DataInvalid,
+                format!("Cannot set default partition spec to unknown spec with id: '{spec_id}'",),
+            )
+                })?
+                .clone();
+        let spec =
+            Arc::unwrap_or_clone(schemaless_spec).bind(self.get_current_schema()?.clone())?;
+        self.metadata.default_spec = Arc::new(spec);
+
+        if self.last_added_spec_id == Some(spec_id) {
+            self.changes.push(TableUpdate::SetDefaultSpec {
+                spec_id: Self::LAST_ADDED,
+            });
+        } else {
+            self.changes.push(TableUpdate::SetDefaultSpec { spec_id });
+        }
+
+        Ok(self)
+    }
+
+    /// Add a partition spec and set it as the default
+    pub fn add_default_partition_spec(self, unbound_spec: UnboundPartitionSpec) -> Result<Self> {
+        self.add_partition_spec(unbound_spec)?
+            .set_default_partition_spec(Self::LAST_ADDED)
+    }
+
+    /// Add a sort order to the table metadata.
+    ///
+    /// The spec is bound eagerly to the current schema and must be valid for it.
+    /// If a schema is added in the same set of changes, the schema should be added first.
+    ///
+    /// Even if `sort_order.order_id` is provided, it may not be used.
+    ///
+    /// # Errors
+    /// - Sort order id to add already exists.
+    /// - Sort order is incompatible with the current schema.
+    pub fn add_sort_order(mut self, sort_order: SortOrder) -> Result<Self> {
+        let new_order_id = self.reuse_or_create_new_sort_id(&sort_order);
+        let sort_order_found = self.metadata.sort_orders.contains_key(&new_order_id);
+
+        if sort_order_found {
+            if self.last_added_order_id != Some(new_order_id) {
+                self.changes.push(TableUpdate::AddSortOrder {
+                    sort_order: sort_order.clone(),
+                });
+                self.last_added_order_id = Some(new_order_id);
+            }
+
+            return Ok(self);
+        }
+
+        // ToDo Discuss: Java builds a fresh spec here:
```

Comment:
We just do this once, after that the schema is evolved using [the `UpdateSchema`](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/UpdateSchema.java), that's implemented by the [SchemaUpdate](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/SchemaUpdate.java). There you pass in the `last-column-id`, and new fields will increment from that.

Reply:
That is actually a very important point, and I think we should fix rust.
Currently in rust we use the `SchemaBuilder` for everything.
To get to it, we use an existing `Schema` and use the `into_builder(self)` method.

I believe we should modify this to `into_builder(self, last_column_id: Option<i64>)` so that users are forced to think about the last_column_id. Most likely they will want to use the `last_column_id` of the metadata they got the `Schema` from.

Reply:
CC @liurenjie1024

Reply:
I'm leaning toward the other way, where this is tracked by Iceberg-rust it self. The last-schema-id is a monotonically increasing ID that keeps track of which field-IDs are already being used.

Reply:
If we do it completely internally, then we should probably not allow or discourage `into_builder` on schema.

I am unsure how an alternative API should look like. @liurenjie1024  some input would be very helpful.
My thoughts:
1. Introduce `TableMetadata.update_schema` that returns the current `SchemaBuilder`, however with a fixed `last-column-id` (set to the `TableMetadata` value). We document that `add_schema` should only be used for schemas modified in this way to ensure that `last_column_id` is correct.
2. If documentation is not enough, we could Introduce a new type `SchemaUpdate` that can only be created by calling `TableMetadata.update_schema`. It would hold the `TableMetadataBuilder` internally and have methods like `add_field` or `remove_field`. Upon build, it would return the `TableMetadataBuilder`.

Reply:
The REST Catalog would use `add_schema` via the `TableUpdate`, which already checks compatibility by validating `last-column-id`.
https://github.com/apache/iceberg-rust/pull/587/files#diff-04f26c83b3c614be6f6d6cfb6c4cefef9e01ec2d31395ac487cdcdff2dbae729R452-R467

Reply:
Hey @liurenjie1024, yes that's a good point. Both in Iceberg-Java and PyIceberg we don't accept the user to pass in a schema directly. Not because we don't trust the user, but mostly because it is easy to make an error. Generally we try to keep field-IDs away from users. To allow evolution of the schema, we have the [`UpdateSchema` API in Python](https://github.com/apache/iceberg-python/blob/2ba86b5c852cb0c8f4b4ba95d54fb5a6bfa00fa2/pyiceberg/table/update/schema.py#L74) and the [`SchemaUpdate` (implements `UpdateSchema`) in Java](https://github.com/apache/iceberg/blob/daa24f9c3a56e18d188097deb2dd79cc991c9a78/core/src/main/java/org/apache/iceberg/SchemaUpdate.java#L48). This API warns the user when they break a schema (for example, changing a `IntType` type to a `StringType`). This API is then used by the `UnionByName` visitor ([python](https://github.com/apache/iceberg-python/blob/2ba86b5c852cb0c8f4b4ba95d54fb5a6bfa00fa2/pyiceberg/table/update/schema.py#L722), [java](https://github.com/apache/iceberg/blob/daa24f9c3a56e18d188097deb2dd79cc991c9a78/core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java#L33)) to make extending of the schema easy. For example, see the [Python docs on Union by Name](https://py.iceberg.apache.org/api/#union-by-name).

Reply:
> I'm just thinking what a rest catalog server should do, since the input may come from arbitratry user

I'd say this depends on the implementation. For example, when you create a new table, the REST catalog often re-assigns the fields (because it doesn't matter). For schema updates, it is more complicated since you need to respect the IDs of the current schema. Keep in mind that even with the APIs, you could break the schema by flipping the [`allow_breaking_changes` flag](https://github.com/apache/iceberg-python/blob/2ba86b5c852cb0c8f4b4ba95d54fb5a6bfa00fa2/pyiceberg/table/update/schema.py#L93) explicitly. So I don't think we can do much here when a user pushes an invalid schema, we can only make sure that they know what they are doing and provide nice APIs to make life easier. I've created #697 and #698 to add those APIs.

Reply:
When we create a new table, we also reassign ids with this builder (and also in Lakekeeper).

I think we agree then that we should have a `SchemaUpdate` in iceberg-rust as well. The issue has already been created by Fokko. Because we don't have the `SchemaUpdate` now, we should keep `into_builder()` method of schemas for now.

REST Catalogs will need to use `add_schema` anyway even if the schema might be incompatible. This is why I need this method to stay public.

Going forward I propose:
* Keep `add_schema` for now as there is no alternative
* Implement #697, add corresponding better methods to the `TableMetadataBuilder` to mutate schema. Discourage the use of `add_schema` in comments, but leave the method public.

What do you think?

Reply:
When adding a word of caution to the method, I'm fine with keeping it public for now.

I know that this is quite late in the process here, but I think it might be complimentary to the builder we're seeing here. For PyIceberg we've taken a different approach, where we apply the `MetadataUpdates` (`AddSchema`, `SetCurrentSchema`, etc) to the metadata, rather than having a builder where you can modify the `TableMetadata`. Code can be found here: https://github.com/apache/iceberg-python/blob/60800d88c7e44fe2ed58a25f1b0fcd5927156adf/pyiceberg/table/update/__init__.py#L215-L494

Reply:
The `TableUpdate` is currently only part of the rest catalog - not of the core spec.
I think we should keep them public unless we go switch to the python approach.

I added a word of warning that we do not check compatibility. Let me know if this is sufficient.
We should extend it with a link to the new method that is implemented in #697 .

Reply:
@liurenjie1024  you are right, I thought it was part of rest.

I am still a bit hesitant to making it private. Let me try to curate my points. None of those are blockers, just want to take them into consideration.
1. I think it's quite useful in tests - good evidence is the rust crate  itself where we use it at a few places. With Lakekeeper we are also using it for tests. Its handy to not require a `TableUpdate` wrapper around the schema that we want to add.
2. There are features that we would need to expose in other ways:
 a) applying multiple `TableUpdates` to a `TableMetadata` in a single shot and obtaining the `TableMetadataBuildResult`, including `changes` and `expired_metadata_logs`. A `Transaction` isn't suitable for this. A `Transaction` works on a `Table`, not `TableMetadata`. `TableMetadata` is much lighter and we as a catalog never actually initialize `Table` - and we don't want to initialize it just to mutate some metadata. `Transaction` also calls the `Catalog` trait in the end - not applying the updates. So its not exposing the Metadata mutating functionality for us.
 This could be solved by adding `pub struct TableUpdates(Vec<TableUpdate>)` or so instead and add an `apply` method to it. There are other options as well of course. Each of them would require another public struct, so we might also stick with the builder.
 b) Initializing `TableMetadata` conveniently including all fields. Currently the only way to do this is via the builder. I don't think it would be ergonomic to have a basic `new()` method on `TableMetadata` and require user to use a completely different mechanism, such as `TableUpdate.apply` to go further (i.e. adding a snapshot or setting the Uuid). `TableCreation` doesn't offer this either in its current state.
 
My baseline is that mutating `TableMetadata` on its own is something that is valuable to external crates, such as our Catalog. So we should offer a public ergonomic interface for it. I believe the builder offers this interface - `Transaction` and `TableUpdate` on its own do not.

Let me know what you think!